core/src/main/java/org/infinispan/remoting/inboundhandler/NonTotalOrderPerCacheInboundInvocationHandler.java

```diff
@@ -27,33 +35,31 @@ public void handle(CacheRpcCommand command, Reply reply, DeliverOrder order) {
          unexpectedDeliverMode(command, order);
       }
       try {
-         boolean onExecutorService = !order.preserveOrder() && command.canBlock();
-         BlockingRunnable runnable;
+         final int commandTopologyId = extractCommandTopologyId(command);
+         if (isCommandSentBeforeFirstTopology(commandTopologyId)) {
+            reply.reply(CacheNotFoundResponse.INSTANCE);
+            return;
+         }
+
+         final boolean waitTransactionalData = command.getCommandId() != StateRequestCommand.COMMAND_ID;
+         final boolean onExecutorService = executeOnExecutorService(order, command);
 
          switch (command.getCommandId()) {
-            case SingleRpcCommand.COMMAND_ID:
-               runnable = createDefaultRunnable(command, reply, extractCommandTopologyId((SingleRpcCommand) command),
-                                                true, onExecutorService);
-               break;
-            case MultipleRpcCommand.COMMAND_ID:
-               runnable = createDefaultRunnable(command, reply, extractCommandTopologyId((MultipleRpcCommand) command),
-                                                true, onExecutorService);
-               break;
-            case StateRequestCommand.COMMAND_ID:
-               // StateRequestCommand is special in that it doesn't need transaction data
-               // In fact, waiting for transaction data could cause a deadlock
-               runnable = createDefaultRunnable(command, reply,
-                     extractCommandTopologyId(((StateRequestCommand) command)), false, onExecutorService);
-               break;
-            default:
-               int commandTopologyId = NO_TOPOLOGY_COMMAND;
-               if (command instanceof TopologyAffectedCommand) {
-                  commandTopologyId = extractCommandTopologyId((TopologyAffectedCommand) command);
-               }
-               runnable = createDefaultRunnable(command, reply, commandTopologyId, true, onExecutorService);
-               break;
+            case PrepareCommand.COMMAND_ID:
+            case VersionedPrepareCommand.COMMAND_ID:
+            case TotalOrderNonVersionedPrepareCommand.COMMAND_ID:
+            case TotalOrderVersionedPrepareCommand.COMMAND_ID:
+            case CommitCommand.COMMAND_ID:
+            case VersionedCommitCommand.COMMAND_ID:
+            case TotalOrderCommitCommand.COMMAND_ID:
+            case TotalOrderVersionedCommitCommand.COMMAND_ID:
+            case RollbackCommand.COMMAND_ID:
+            case TotalOrderRollbackCommand.COMMAND_ID:
+            case LockControlCommand.COMMAND_ID:
+               throw new IllegalArgumentException("Illegal Command ID: " + command.getCommandId() + ". Transactional commands are not allowed.");
```

Comment:
Maybe it would have been simpler to check for `SingleRpcCommand` and `MultipleRpcCommand` instead?